### PR TITLE
THRIFT-4372 Pipe write operations across a network are limited to 65,…

### PR DESCRIPTION
**Pull request by Jens-G**
…535 bytes per write

Client: C#
Patch: Jens Geyer

### CODE REVIEW

File: lib/csharp/src/Transport/TNamedPipeClientTransport.cs [hunk 1]
```diff
@@ -88,7 +89,18 @@ public override void Write(byte[] buf, int off, int len)
                 throw new TTransportException(TTransportException.ExceptionType.NotOpen);
             }
 
-            client.Write(buf, off, len);
+            // if necessary, send the data in chunks
+            // there's a system limit around 0x10000 bytes that we hit otherwise
+            // MSDN: "Pipe write operations across a network are limited to 65,535 bytes per write. For more information regarding pipes, see the Remarks section."
+            var nBytes = Math.Min(len, 15 * 4096);  // 16 would exceed the limit
```
Comment (jeking3) [inline lib/csharp/src/Transport/TNamedPipeClientTransport.cs]: Why not actually use `(2^16)-1` which is the limit?

File: lib/csharp/src/Transport/TNamedPipeClientTransport.cs [hunk 2]
```diff
@@ -88,7 +89,18 @@ public override void Write(byte[] buf, int off, int len)
...
+            while (nBytes > 0)
+            {
+                client.Write(buf, off, nBytes);
+
+                off += nBytes;
+                len -= nBytes;
+                nBytes = Math.Min(len, nBytes);
```
Comment (jeking3) [inline lib/csharp/src/Transport/TNamedPipeClientTransport.cs]: Shouldn't this be the same calculation as the one outside the loop?  Technically it doesn't have to be, however I found this confusing to read.

File: lib/csharp/src/Transport/TNamedPipeServerTransport.cs [hunk 3]
```diff
@@ -239,40 +239,51 @@ public override void Write(byte[] buf, int off, int len)
                     throw new TTransportException(TTransportException.ExceptionType.NotOpen);
                 }
 
-                if (asyncMode)
+                // if necessary, send the data in chunks
+                // there's a system limit around 0x10000 bytes that we hit otherwise
+                // MSDN: "Pipe write operations across a network are limited to 65,535 bytes per write. For more information regarding pipes, see the Remarks section."
+                var nBytes = Math.Min(len, 15 * 4096);  // 16 would exceed the limit
```
Comment (jeking3) [inline lib/csharp/src/Transport/TNamedPipeServerTransport.cs]: Why not actually use (2^16)-1 which is the limit?

File: lib/csharp/src/Transport/TNamedPipeServerTransport.cs [hunk 4]
```diff
@@ -239,40 +239,51 @@ public override void Write(byte[] buf, int off, int len)
...
+                    else
+                    {
+                        stream.Write(buf, off, nBytes);
+                    }
 
-                    if (eOuter != null)
-                        throw eOuter; // rethrow exception
+                    off += nBytes;
+                    len -= nBytes;
+                    nBytes = Math.Min(len, nBytes);
```
Comment (jeking3) [inline lib/csharp/src/Transport/TNamedPipeServerTransport.cs]: Shouldn't this be the same calculation as the one outside the loop?  Technically it doesn't have to be, however I found this confusing to read.

**Comment by Jens-G** [comment]
> Why not actually use (2^16)-1 which is the limit?

Several reasons. First, aligned memory access is always faster. If we subtract 1 byte, we get the worst case. Next, at least on Windows a number of system resources use multiples of 4096, so it is probably not a bad idea to follow that model. That's why I picked this particular value.

**Comment by jeking3** [comment]
Makes sense, thanks for the info.  Every little bit helps us all improve what we do next.

**Pull request closed without merging**